plugins/elantp/fu-elantp-hid-haptic-device.c

```diff
@@ -731,14 +695,65 @@ fu_elantp_hid_haptic_device_write_firmware(FuDevice *device,
 				    error_iapctrl->message);
 			return FALSE;
 		}
-		retry_cnt = 0;
 
 		/* update progress */
-		checksum += csum_tmp;
-		fu_progress_set_percentage_full(fu_progress_get_child(progress),
-						(gsize)i + 1,
-						(gsize)chunks->len + 1);
+		helper->checksum += csum_tmp;
+		fu_progress_step_done(helper->progress);
 	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_elantp_hid_haptic_device_write_firmware(FuDevice *device,
+					   FuFirmware *firmware,
+					   FuProgress *progress,
+					   FwupdInstallFlags flags,
+					   GError **error)
+{
+	FuElantpHidDevice *parent;
+	FuElantpHidHapticDevice *self = FU_ELANTP_HID_HAPTIC_DEVICE(device);
+	guint16 checksum_device = 0;
+	const gchar *fw_ver;
+	const gchar *fw_ver_device;
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GError) error_local = NULL;
+	FuElantpHaptictpWaitFlashEEPROMChecksumHelper helper = {0x0};
+	FuElantpHaptictpWriteHelper helper_write = {0x0};
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 5, "detach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 80, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_VERIFY, 10, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 5, NULL);
+
+	/* simple image */
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	/* use parent */
+	parent = fu_elantp_haptic_device_get_parent(device, error);
+	if (parent == NULL)
+		return FALSE;
+
+	/* detach */
+	if (!fu_elantp_hid_haptic_device_detach(device, fu_progress_get_child(progress), error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* write each block */
+	helper_write.fw = fw;
+	helper_write.progress = fu_progress_get_child(progress);
+	if (!fu_device_retry_full(device,
+				  fu_elantp_hid_haptic_device_write_chunks_cb,
+				  3,
+				  100,
+				  &helper_write,
+				  error))
+		return FALSE;
```

Comment:
It is retry 3 times for only current page, not all page retry again.

Reply:
Understood; can you try the fixed version please.

Reply:
It's works, but some tp firmware have "tp busy issue, when write the page, ", it's hard to reproduce.